Cannot build examples(amqp_ssl_connect.exe) on Windows/mingw32/gcc: undefined reference to `InitializeSRWLock'

Hello,

After building examples without SSL/TLS support (https://github.com/alanxz/rabbitmq-c/issues/485) I downloaded latest OpenSSL sources from github and built them according to the steps describend on http://wiki.qt.io/Compiling_OpenSSL_with_MinGW.

However, building examples(amqp_ssl_connect.exe) fails with:

```cmd
>C:\CMake\bin\cmake.exe --version
cmake version 3.10.3

>mingw32-make.exe -v
GNU Make 4.1
Built for i686-w64-mingw32

>gcc.exe --version
gcc.exe (i686-posix-dwarf-rev0, Built by MinGW-W64 project) 5.3.0

>c:\CMake\bin\cmake.exe -DCMAKE_INSTALL_PREFIX=E:\eegorov\amqp\rabbitmq-c\bin -DBUILD_SHARED_LIBS=OFF -DBUILD_STATIC_LIBS=ON -DENABLE_SSL_SUPPORT=ON -DBUILD_TESTS=OFF -DBUILD_EXAMPLES=ON -DCMAKE_PREFIX_PATH=E:\eegorov\amqp\openssl\dist;C:\Qt\Qt5.8.0\Tools\mingw530_32\i686-w64-mingw32 -DCMAKE_BUILD_TYPE=Release -G "MinGW Makefiles" ..
-- The C compiler identification is GNU 5.3.0
-- Check for working C compiler: C:/Qt/Qt5.8.0/Tools/mingw530_32/bin/gcc.exe
-- Check for working C compiler: C:/Qt/Qt5.8.0/Tools/mingw530_32/bin/gcc.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Found C inline keyword: inline
-- Looking for poll
-- Looking for poll - not found
-- Looking for clock_gettime in rt
-- Looking for clock_gettime in rt - not found
-- Looking for posix_spawnp in rt
-- Looking for posix_spawnp in rt - not found
-- Found OpenSSL: E:/eegorov/amqp/openssl/dist/lib/libcrypto.dll.a (found suitable version "1.1.1", minimum required is "0.9.8")
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE
-- Performing Test HAVE_GNU90
-- Performing Test HAVE_GNU90 - Success
-- Could NOT find POPT (missing: POPT_INCLUDE_DIR POPT_LIBRARY) (found version "")
-- Could NOT find XMLTO (missing: XMLTO_EXECUTABLE)
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE)
-- Building rabbitmq as a shared library - no
-- Building rabbitmq as a static library - yes
-- Configuring done
-- Generating done
-- Build files have been written to: E:/eegorov/amqp/rabbitmq-c/dist

>c:\CMake\bin\cmake.exe --build . --config Release --target install
Scanning dependencies of target rabbitmq-static
[  1%] Building C object librabbitmq/CMakeFiles/rabbitmq-static.dir/amqp_framing.c.obj
[  3%] Building C object librabbitmq/CMakeFiles/rabbitmq-static.dir/amqp_api.c.obj
[  5%] Building C object librabbitmq/CMakeFiles/rabbitmq-static.dir/amqp_connection.c.obj
[  6%] Building C object librabbitmq/CMakeFiles/rabbitmq-static.dir/amqp_mem.c.obj
/* . . . */
[ 90%] Building C object examples/CMakeFiles/amqp_listenq.dir/utils.c.obj
[ 91%] Building C object examples/CMakeFiles/amqp_listenq.dir/win32/platform_utils.c.obj
[ 93%] Linking C executable amqp_listenq.exe
[ 93%] Built target amqp_listenq
Scanning dependencies of target amqp_ssl_connect
[ 95%] Building C object examples/CMakeFiles/amqp_ssl_connect.dir/amqp_ssl_connect.c.obj
[ 96%] Building C object examples/CMakeFiles/amqp_ssl_connect.dir/utils.c.obj
[ 98%] Building C object examples/CMakeFiles/amqp_ssl_connect.dir/win32/platform_utils.c.obj
[100%] Linking C executable amqp_ssl_connect.exe
../librabbitmq/liblibrabbitmq.4.a(threads.c.obj):threads.c:(.text+0x24): undefined reference to `InitializeSRWLock'
../librabbitmq/liblibrabbitmq.4.a(threads.c.obj):threads.c:(.text+0x44): undefined reference to `AcquireSRWLockExclusive'
../librabbitmq/liblibrabbitmq.4.a(threads.c.obj):threads.c:(.text+0x64): undefined reference to `ReleaseSRWLockExclusive'
collect2.exe: error: ld returned 1 exit status
examples\CMakeFiles\amqp_ssl_connect.dir\build.make:153: recipe for target 'examples/amqp_ssl_connect.exe' failed
mingw32-make.exe[2]: *** [examples/amqp_ssl_connect.exe] Error 1
CMakeFiles\Makefile2:529: recipe for target 'examples/CMakeFiles/amqp_ssl_connect.dir/all' failed
mingw32-make.exe[1]: *** [examples/CMakeFiles/amqp_ssl_connect.dir/all] Error 2
Makefile:128: recipe for target 'all' failed
mingw32-make.exe: *** [all] Error 2
```

According to https://msdn.microsoft.com/ru-ru/library/windows/desktop/ms683483(v=vs.85).aspx they are in Kernel32.dll library, so I tried to modify `examples\CMakeLists.txt` with both:
```cmake
set (amqp_ssl_connect_LIBS ${RMQ_LIBRARY_TARGET} "C:/Qt/Qt5.8.0/Tools/mingw530_32/i686-w64-mingw32/lib/libkernel32.a")
target_link_libraries(amqp_ssl_connect ${amqp_ssl_connect_LIBS})
```
and
```cmake
set (amqp_ssl_connect_LIBS ${RMQ_LIBRARY_TARGET} kernel32)
target_link_libraries(amqp_ssl_connect ${amqp_ssl_connect_LIBS})
```
but without any luck. 

I'm newbie in CMake that's why I decided to double-check it: found with `mingw32-make.exe -n` execution string and run:
```cmd
>cd /d E:\eegorov\amqp\rabbitmq-c\dist\examples && C:\CMake\bin\cmake.exe -E cmake_link_script CMakeFiles\amqp_ssl_connect.dir\link.txt --verbose=
../librabbitmq/liblibrabbitmq.4.a(threads.c.obj):threads.c:(.text+0x24): undefined reference to `InitializeSRWLock'
../librabbitmq/liblibrabbitmq.4.a(threads.c.obj):threads.c:(.text+0x44): undefined reference to `AcquireSRWLockExclusive'
../librabbitmq/liblibrabbitmq.4.a(threads.c.obj):threads.c:(.text+0x64): undefined reference to `ReleaseSRWLockExclusive'
collect2.exe: error: ld returned 1 exit status
```
=> content of `CMakeFiles\amqp_ssl_connect.dir\link.txt` is:
```
C:\CMake\bin\cmake.exe -E remove -f CMakeFiles\amqp_ssl_connect.dir/objects.a
C:\Qt\Qt5.8.0\Tools\mingw530_32\i686-w64-mingw32\bin\ar.exe cr CMakeFiles\amqp_ssl_connect.dir/objects.a @CMakeFiles\amqp_ssl_connect.dir\objects1.rsp
C:\Qt\Qt5.8.0\Tools\mingw530_32\bin\gcc.exe -Wall -Wextra -Wstrict-prototypes -Wno-unused-function -fno-common -fvisibility=hidden  -std=gnu90 -O3 -DNDEBUG   -Wl,--whole-archive CMakeFiles\amqp_ssl_connect.dir/objects.a -Wl,--no-whole-archive  -o amqp_ssl_connect.exe -Wl,--out-implib,libamqp_ssl_connect.dll.a -Wl,--major-image-version,0,--minor-image-version,0 @CMakeFiles\amqp_ssl_connect.dir\linklibs.rsp
```
=> finally, in `CMakeFiles\amqp_ssl_connect.dir\linklibs.rsp` we have:
```
../librabbitmq/liblibrabbitmq.4.a -Wl,-Bstatic -lkernel32 -Wl,-Bdynamic E:/eegorov/amqp/openssl/dist/lib/libssl.dll.a E:/eegorov/amqp/openssl/dist/lib/libcrypto.dll.a -lws2_32 -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32
```
So kernel32 is there, even twice (probably for ordering dependencies). Can anybody give me a hint on how to solve this linking issue please? Or, if you successfully built SSL-enabled examples on Windows - could you share your environment/toolchain details please?

@sigiesec Thank you, you are right! For some weird reason (for me only, ofc), `_WIN32_WINNT` is not `_WIN32_WINNT_WIN7` on my Win7 machine. I've added to `librabbitmq/win32/threads.h`:
```diff
diff --git a/librabbitmq/win32/threads.h b/librabbitmq/win32/threads.h
index 3ae571f..3bbc069 100644
--- a/librabbitmq/win32/threads.h
+++ b/librabbitmq/win32/threads.h
@@ -34,6 +34,9 @@
 #ifndef WIN32_LEAN_AND_MEAN
 #define WIN32_LEAN_AND_MEAN
 #endif
+
+#define _WIN32_WINNT _WIN32_WINNT_WIN7
+
 #include <windows.h>

 typedef SRWLOCK pthread_mutex_t;
```
and sucessfully rebuilt project! Just compiler complained a bit about it with:
```
[ 16%] Building C object librabbitmq/CMakeFiles/rabbitmq-static.dir/amqp_consumer.c.obj
[ 18%] Building C object librabbitmq/CMakeFiles/rabbitmq-static.dir/amqp_openssl.c.obj
In file included from E:\eegorov\amqp\rabbitmq-c\librabbitmq\amqp_openssl.c:37:0:
E:/eegorov/amqp/rabbitmq-c/librabbitmq/win32/threads.h:38:0: warning: "_WIN32_WINNT" redefined
 #define _WIN32_WINNT _WIN32_WINNT_WIN7
 ^
```

--- a/librabbitmq/win32/threads.h
+++ b/librabbitmq/win32/threads.h
@@ -27,7 +27,10 @@
 #ifndef AMQP_THREAD_H
 #define AMQP_THREAD_H
 
-#ifndef WINVER
+#if !defined(WINVER) || defined(__MINGW32__) || defined(__MINGW64__)
+#ifdef WINVER
+#undef WINVER
+#endif
 /* Windows Vista or newer */
 #define WINVER 0x0600
 #endif

